### DISCUSSION TITLE
Limited mode

### DISCUSSION THREAD

**Discussion by sichtbeton** [SynthstromAudible/DelugeFirmware#605] [Ideas]
suggestion: a limited mode of (maybe 8?) fixed tracks that will always work. maybe kits are limited to a certain number of rows too. this way one could work with a fixed expectation of what the device is able to handle. makes a serious difference knowing what you are up against instead of guesswork. features could be:

- 8 tracks maximum (maybe. 6 notes max per voice)
- can be samples or synths
- all parameters are moddablle / possible (if too CPU heavy, forbid them)
- grain or other heavy stuff is off

this would ensure the following:

- know what you have at your disposal and work with that
- no more guessing why it cuts samples or long release synths
- confidence, that these 8 tracks really work (important for live)
- would fit the 8 rows of the Deluge pad grid

It would kind of replicate the OP-Z, octatrack and others, that also have limits but within those limits they just work.

what do you guys think?

**Comment by jamiefaye** [comment]
One idea is to have what I call "temporal regression tests". This would be a test suite of songs that would be expected to be rendered to audio within certain time constraints on aspects of the process. If something that used to be within spec drifts-out, we would then know and could find compensations to restore performance.

We could choose these tests and their limits so as to have something like what you describe in terms of number of tracks, notes, effects, etc.

**Comment by entzmingerc** [comment]
Brainstormed a bit on the discord! Thought I'd copy some stuff over here. 

I think if we built some test songs on Official Firmware then we'd have an idea of what is our "expected number of sample voices in a kit" supposed to be or any other limit. Building off Jamie's idea a bit more:

**SongName**: test-kits
**Test**: 1 sample voice per row, 8 rows of kits, 8 kit clips in a section, trigger one sample, then 4 samples, then all 8 samples, then multiple samples across each kit, progressively increasing the number of samples being triggered simultaneously across all 8 clips
**Results**: (what does official do? It probably doesn't trigger all 64 samples at once, but when does culling start to happen? What about long samples vs short samples?)

I think we want to "match official's performance" but haven't got a number for it, and the specific number depends on what you're doing. If we knew what we're getting from official more clearly, then we'd know what to expect on community so when we test things we're not guessing if it's a problem or not. 

I'm not sure how to figure out "at what point do the kits start culling voices" but I think if we make some tests songs like this the devs have the tools to see that data.
